include/graaflib/algorithm/cycle_detection.tpp

```diff
@@ -0,0 +1,96 @@
+#pragma once
+
+#include <algorithm>
+#include <unordered_map>
+
+namespace graaf::algorithm {
+
+namespace detail {
+
+enum vertex_color{WHITE, GREY, BLACK};
+
+template <typename V, typename E, graph_spec S>
+bool do_dfs_directed(const graph<V, E, S>& graph,
+                     std::unordered_map<vertex_id_t, int>& colored_vertices,
+                     vertex_id_t current) {
+  colored_vertices[current] = vertex_color::GREY;
+
+  for (const auto& neighbour_vertex : graph.get_neighbors(current)) {
+    if (colored_vertices[neighbour_vertex] == vertex_color::WHITE) {
+        if (do_dfs_directed(graph, colored_vertices, neighbour_vertex))
+            return true;
+    } 
+    else if (colored_vertices[neighbour_vertex] == vertex_color::GREY) {
+      return true;
+    }
+  }
+
+  colored_vertices[current] = vertex_color::BLACK;
+  return false;
+}
+
+template<typename V, typename E, graph_spec S>
+bool do_dfs_undirected(const graph<V,E,S> & graph, 
+    std::unordered_map<vertex_id_t, bool> & visited_vertices,
+    std::unordered_map<vertex_id_t, vertex_id_t> & parent_vertices,
+    vertex_id_t parent_vertex, vertex_id_t current) {
+
+    visited_vertices[current] = true;
+
+    for (const auto& neighbour_vertex : graph.get_neighbors(current)) {
+        if (neighbour_vertex == parent_vertex) 
+            continue;
+
+        if (visited_vertices[neighbour_vertex]) 
+            return true;
+
+        parent_vertices[neighbour_vertex] = parent_vertex;
+
+        if (do_dfs_undirected(graph, visited_vertices, parent_vertices, neighbour_vertex,
+                            parent_vertices[neighbour_vertex])) {
+                                     return true;
+        }
+    }
+
+    return false;
+}
+
+}  // namespace detail
+
+template <typename V, typename E, graph_spec S>
+bool has_cycle(const graph<V, E, S>& graph) {
+
+
+  if (graph.is_directed()) {
+        std::unordered_map<vertex_id_t, int> colored_vertices{};
+
+        for (const auto& vertex : graph.get_vertices()) {
+            if (!colored_vertices.contains(vertex.first) &&
+                detail::do_dfs_directed(graph, colored_vertices, vertex.first)) {
+                        return true;
+            }
+        }
+
+    return false;
+  }
+
+  if (graph.is_undirected()) {
```

Comment:
[minor]: in the case of an undirected graph, we can return early if the number of edges is equal to or greater than the number of vertices. What do you think of adding this here?
```suggestion
  if (graph.is_undirected()) {
       if (graph.edge_count() >= graph.vertex_count()) {
         return true;
       }
```

Reply:
Good suggestion, since it's undirected graph and we cannot add edge with vertices that out of range this one will be correct!